Dont block the agent if we can't send logs

iksaif opened the pull request:
### What does this PR do?

Currently if the Senders are not able to consume the messages of the
pipelines (for example, because they can't establish a connection) it
is impossible to stop the logging agent. Even worse, it is impossible
to stop the inputs too because the channels in-between components have
a fixed size.


### Additional Notes

This is based heavily on context.Context to allow us to have proper
composite timeouts / cancellations. This is the recommended way to do
that in go which is also already used in other parts of the agent.

This commit also changes the following things:
- Ads a DestinationContext that is restartable and allows to "unclog" the destinations when necessary
- Destination starts using context.Context in order to have proper cancelation timeouts.
- Add some cleanups in a few `Stop()` functions to make unit testing easier
- Add `agent_test.go`, `sender_test.go` and `connection_manager_test.go` with basic tests (with working and non-working backends)
- Add `logs_config.stop_grace_period`, both because this has an effect on the logs that could be lost on restart (for tcp/udp) and because it makes unit testing easier.

iksaif commented on pkg/logs/sender/sender.go:
```diff
@@ -59,20 +84,46 @@ func (s *Sender) run() {
+		// when we are stopping we are still reading the messages from inputChan (so
+		// that the rest of the pipeline finishes up cleanly) but we are not trying to
+		// connect anymore.
+		select {
```
move that in run()

iksaif replied:
done

iksaif commented on pkg/logs/sender/sender.go:
```diff
@@ -59,20 +84,46 @@ func (s *Sender) run() {
+		}
+
+		// If we've been cancelled we don't want to forward messages anymore.
+		if err == context.Canceled {
```
Remove

iksaif replied:
done

iksaif commented on pkg/logs/sender/sender.go:
```diff
@@ -59,20 +84,46 @@ func (s *Sender) run() {
+		// This happens when the connection attempt has been interrupted.
 		if s.conn == nil {
-			s.conn = s.connManager.NewConnection() // blocks until a new conn is ready
+			continue
```
break (on err != nill or s.conn == nill)

iksaif replied:
done

iksaif commented on pkg/logs/sender/sender.go:
```diff
@@ -36,14 +41,34 @@ func New(inputChan, outputChan chan message.Message, connManager *ConnectionMana
 
 // Start starts the Sender
 func (s *Sender) Start() {
+	s.ctx, s.cancel = context.WithCancel(context.Background())
```
done

ajacquemot reviewed (commented):
Left few comments, this looks great overall, happy to see new go-patterns :-)

iksaif commented on pkg/logs/sender/sender.go:
```diff
@@ -60,19 +88,28 @@ func (s *Sender) run() {
 		if err != nil {
-			log.Error("can't send payload: ", payload, err)
+			log.Error("can't serialize payload: ", payload, err)
 			continue
```
If we return here we lose the current message forever

ajacquemot replied:
Yep but the delimiting operation is idempotent so if you fail once, you'll fail forever.

iksaif commented on pkg/logs/sender/sender.go:
```diff
@@ -52,6 +73,13 @@ func (s *Sender) run() {
+		default:
+		}
+
 		s.wireMessage(payload)
```
done

iksaif commented on pkg/logs/logs.go:
```diff
@@ -59,10 +62,17 @@ func Start() error {
-	if isRunning {
-		log.Info("Stopping logs-agent")
+	log.Info("Stopping logs-agent...")
+	if agent != nil {
```
What do you mean by "A Client" ?

ajacquemot replied:
a user of this method

iksaif committed:
logs/input: fix data race in tailer

Also fixes a similar race with shouldStop (that hasn't been triggered yet)

```
WARNING: DATA RACE
Write at 0x00c0420dc2b9 by goroutine 46:
  github.com/DataDog/datadog-agent/pkg/logs/input/file.(*Tailer).Stop()
      C:/gopath/src/github.com/DataDog/datadog-agent/pkg/logs/input/file/tailer.go:96 +0x47
  github.com/DataDog/datadog-agent/pkg/logs/restart.(*parallelStopper).Stop.func1()
      C:/gopath/src/github.com/DataDog/datadog-agent/pkg/logs/restart/parallel_stop.go:35 +0x45
Previous read at 0x00c0420dc2b9 by goroutine 29:
  github.com/DataDog/datadog-agent/pkg/logs/input/file.(*Tailer).forwardMessages()
      C:/gopath/src/github.com/DataDog/datadog-agent/pkg/logs/input/file/tailer.go:177 +0x1d4
Goroutine 46 (running) created at:
  github.com/DataDog/datadog-agent/pkg/logs/restart.(*parallelStopper).Stop()
      C:/gopath/src/github.com/DataDog/datadog-agent/pkg/logs/restart/parallel_stop.go:34 +0x108
  github.com/DataDog/datadog-agent/pkg/logs/input/file.(*Scanner).cleanup()
      C:/gopath/src/github.com/DataDog/datadog-agent/pkg/logs/input/file/scanner.go:89 +0x213
  github.com/DataDog/datadog-agent/pkg/logs/input/file.(*Scanner).Stop()
      C:/gopath/src/github.com/DataDog/datadog-agent/pkg/logs/input/file/scanner.go:60 +0x6d
  github.com/DataDog/datadog-agent/pkg/logs/restart.(*parallelStopper).Stop.func1()
      C:/gopath/src/github.com/DataDog/datadog-agent/pkg/logs/restart/parallel_stop.go:35 +0x45
Goroutine 29 (running) created at:
  github.com/DataDog/datadog-agent/pkg/logs/input/file.(*Tailer).Start()
      C:/gopath/src/github.com/DataDog/datadog-agent/pkg/logs/input/file/tailer.go:81 +0x189
  github.com/DataDog/datadog-agent/pkg/logs/input/file.(*Scanner).startNewTailer()
      C:/gopath/src/github.com/DataDog/datadog-agent/pkg/logs/input/file/scanner.go:188 +0x278
  github.com/DataDog/datadog-agent/pkg/logs/input/file.(*Scanner).launchTailers()
      C:/gopath/src/github.com/DataDog/datadog-agent/pkg/logs/input/file/scanner.go:174 +0x244
  github.com/DataDog/datadog-agent/pkg/logs/input/file.(*Scanner).run()
      C:/gopath/src/github.com/DataDog/datadog-agent/pkg/logs/input/file/scanner.go:71 +0x319
```

achntrl commented on pkg/logs/sender/sender.go:
```diff
@@ -60,19 +90,28 @@ func (s *Sender) run() {
+			s.conn, _ = s.connManager.NewConnection(s.ctx)
+		}
+
+		// This happens when the connection attempt has been interrupted (whe stopping).
```
nits: `whe`

achntrl commented on pkg/logs/agent_test.go:
```diff
@@ -0,0 +1,99 @@
+
+import (
+	"fmt"
+	"github.com/DataDog/datadog-agent/pkg/logs/sender/mock"
```
nits: this import should be grouped with other `github.com/DataDog/datadog-agent/pkg/logs/` imports

iksaif replied:
fixed

achntrl commented on pkg/logs/sender/connection_manager_test.go:
```diff
@@ -0,0 +1,60 @@
+package sender
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/logs/sender/mock"
```
nits: this import should be grouped with other `github.com/DataDog/datadog-agent/pkg/logs/` imports

iksaif replied:
fixed

achntrl commented on pkg/logs/sender/sender.go:
```diff
@@ -6,21 +6,26 @@
 
 import (
+	"context"
+	"github.com/DataDog/datadog-agent/pkg/logs/config"
```
nits: this import should be grouped with other `github.com/DataDog/datadog-agent/pkg/logs/` imports

iksaif commented:
Blocked on https://github.com/DataDog/datadog-agent/pull/2381, will address comments after.

iksaif commented:
@ajacquemot tell me if you have started merging this, if not I probably will have the time to do it today

ajacquemot commented:
Did not start yet @iksaif, don't hesitate to reach out to me if needed, thanks.

iksaif committed:
pkg/logs: unblock pipelines when we have no working sender

iksaif commented:
Rebased, I'm still not happy with a few things.

iksaif committed:
pkg/logs: keep the context on Stop() in DestinationsContext

ajacquemot commented on pkg/logs/sender/connection_manager.go:
```diff
@@ -70,29 +81,36 @@ func (cm *ConnectionManager) NewConnection() net.Conn {
-			conn, err = net.DialTimeout("tcp", cm.address(), timeout)
+			var dialer net.Dialer
+			dctx, cancel := context.WithTimeout(ctx, connectionTimeout)
+			defer cancel()
```
why do we defer `cancel` here ?

ajacquemot replied:
👍

ajacquemot commented on pkg/logs/sender/destination.go:
```diff
@@ -32,26 +33,45 @@ func (e *FramingError) Error() string {
-	conn        net.Conn
+	prefixer  Prefixer
+	delimiter Delimiter
+	/* TODO: Merge ConnectionManager into Destination (think about GRPC integration before) */
```
I think this is not likely to happen now and if this happens, we'll use core agent `Forwarder` I think

iksaif replied:
removed

ajacquemot commented on pkg/logs/sender/destination.go:
```diff
@@ -32,26 +33,45 @@ func (e *FramingError) Error() string {
+		var err error
+
+		// If we have a context, use it, this will allow early cancellation.
+		if d.destinationsContext != nil {
```
as discussed IRL, it seemed that you had a better idea to tackle this.

iksaif replied:
This was for `d.destinationsContext.Context() = nil` (after a `Stop()`), which I did.
This check is still necessary for make the `d.destinationsContext` optional (unit tests)

ajacquemot commented on pkg/logs/sender/destinations_context.go:
```diff
@@ -0,0 +1,51 @@
+}
+
+// Start creates a context that will be cancelled on Stop()
+func (sm *DestinationsContext) Start() {
```
what does `sm` stand for ? can you change this to `dc` please ?

iksaif replied:
it stands for "bad copy paste"

ajacquemot commented on pkg/logs/sender/sender.go:
```diff
@@ -53,8 +54,14 @@ func (s *Sender) run() {
+		// this call is blocking until payload is sent (or the connection destination context cancelled)
+		err := s.destinations.Main.Send(payload)
 		if err != nil {
+			if err == context.Canceled {
```
Wondering whether we could merge this check with the switch right bellow.

iksaif replied:
did not find how to do that :/

ajacquemot reviewed (commented):
Thanks for rebasing on top of my work, overall this looks good to me, I left few comments.

iksaif committed:
logs: small cleanups and typos

iksaif commented:
Fixed the comments.

ajacquemot commented on pkg/logs/input/file/tailer.go:
```diff
@@ -42,8 +42,8 @@ type Tailer struct {
 	closeTimeout  time.Duration
-	shouldStop    bool
-	didFileRotate bool
+	shouldStop    int32
```
that's a pity that bool can not be atomic ..

iksaif committed:
logs: run gofmt

ajacquemot commented on pkg/logs/sender/destination.go:
```diff
@@ -32,26 +33,40 @@ func (e *FramingError) Error() string {
+		var err error
+
+		// If we have a context, use it, this will allow early cancellation.
+		if d.destinationsContext != nil {
```
I am not a huge fan of this, I think a sender should always have a `destinationsContext `

iksaif replied:
So I looked again and .. The relationships between destinations senders and destinations context are quite complicated.

Currently, if there is a bug and destinationsContext isn't set properly, we just fail to cancel the destinations on shutdown. If I remove the check, we will crash. Can we keep the safer way for now, and re-evaluate this later ?

ajacquemot replied:
👍

ajacquemot reviewed (dismissed):
I added one last comment, I think we're good to get this merged, I'd like someone else from team intake to have a look though.
Thanks

iksaif commented:
Rebased :)

achntrl commented on pkg/logs/sender/sender.go:
```diff
@@ -53,8 +54,14 @@ func (s *Sender) run() {
 		if err != nil {
+			if err == context.Canceled {
+				// the context was cancelled, agent is stopping non-gracefully.
+				// drop the message
```
Should we log that we're dropping message ?

achntrl reviewed (commented):
Very interesting pattern with the `Context` objets

hush-hush commented on pkg/config/config.go:
```diff
@@ -282,6 +282,7 @@ func init() {
 	BindEnvAndSetDefault("logs_config.dd_port", 10516)
 	BindEnvAndSetDefault("logs_config.dev_mode_use_proto", true)
 	BindEnvAndSetDefault("logs_config.dd_url_443", "agent-443-intake.logs.datadoghq.com")
+	BindEnvAndSetDefault("logs_config.stop_grace_period", "30s")
```
All duration in the config are express without the `s` and use int. Would be nice to do the same for consistency.

iksaif replied:
Does that mean that we should never use `GetDuration()` for config values (see https://github.com/spf13/viper#getting-values-from-viper) ? 

There are two existing usages for `GetDuration()` which look wrong since they multiply the result by time.Seconds)

iksaif replied:
Discussed on slack, let's use `GetInt` for now and then convert everything to a wrapped `GetDuration`

iksaif committed:
config: logs_config.stop_grace_period is now an Int

For consistency, I'll convert all Durations to actual Durations
after this PR.

achntrl commented on pkg/logs/sender/sender.go:
```diff
@@ -64,7 +64,7 @@ func (s *Sender) send(payload *message.Message) {
 			switch err.(type) {
 			case *FramingError:
-				// the message can not be framed properly,
+				// the message can notx be framed properly,
```
:(

iksaif committed:
logs/sender: remove spurious x
